### ISSUE TITLE
SDL fullscreen broken

### ISSUE THREAD

**Issue by stsp** [stsp/dosemu2#350]
Ctrl-Att-f switch doesn't work as expected.

**Comment by stsp** [comment]
Works fine on fedora-25, but not on ubuntu.

**Comment by stsp** [comment]
Guys, could please everyone try Ctrl-Alt-f
and write the results here?
For me works fine on SDL2-2.0.5-3.fc25.x86_64
but not on ubuntu's SDL2 (don't remember the
version right now). But this may also be WM-related.

**Comment by andrewbird** [comment]
I never used it before, but seems to work fine on Lubuntu 14.04 32bit
~~~
$ apt search libsdl2
libsdl2-2.0-0/xenial,now 2.0.4+dfsg1-2ubuntu2 i386 [installed,automatic]

$ apt search lxde | grep installed
libmenu-cache-bin/xenial,now 1.0.1-1build1 i386 [installed,automatic]
libmenu-cache3/xenial,now 1.0.1-1build1 i386 [installed,automatic]
lxappearance/xenial,now 0.6.2-1 i386 [installed]
lxappearance-obconf/xenial,now 0.2.3-1 i386 [installed]
lxde-common/xenial,now 0.99.1-1ubuntu1 all [installed,automatic]
lxde-core/xenial,now 7ubuntu1 all [installed,automatic]
lxinput/xenial,now 0.3.5-1 i386 [installed]
lxlauncher/xenial,now 0.2.5-1 i386 [installed]
lxmenu-data/xenial,now 0.1.5-1 all [installed,automatic]
lxpanel/xenial,now 0.8.2-1ubuntu2 i386 [installed]
lxpanel-data/xenial,now 0.8.2-1ubuntu2 all [installed,automatic]
lxrandr/xenial,now 0.3.1-1 i386 [installed]
lxsession/xenial,now 0.5.2-0ubuntu3 i386 [installed]
lxsession-logout/xenial,now 0.5.2-0ubuntu3 i386 [installed]
lxshortcut/xenial-updates,now 1.2.4-1ubuntu1.1 i386 [installed]
lxtask/xenial,now 0.1.7-1 i386 [installed]
lxterminal/xenial,now 0.2.0-1 i386 [installed]

$ wmctrl -m
Name: Openbox
Class: 
PID: N/A
Window manager's "showing the desktop" mode: OFF

openbox/xenial,now 3.6.1-1ubuntu2 i386 [installed]
~~~

**Comment by stsp** [comment]
Thanks.
I had it broken on ubuntu-16.04.
So if it is not broken for others, maybe this
is not a stopper for release. OTOH I still can't
make the packages without a working tar on
fedora, so there is nothing to rush for...

**Comment by andrewbird** [comment]
Of course my Lubuntu is probably not representative of vanilla Ubuntu 16.04 as it's not running the same desktop etc.

**Comment by bolle732** [comment]
I don't have problems on Ubuntu 16.04 64-bit with switching to full screen in SDL mode. This Ubuntu is all standard with Unity desktop.
```
$ dpkg -l "*sdl2*"
||/ Name                       Version            Architecture       Description
+++-==========================-==================-==================-=========================================================
ii  libsdl2-2.0-0:amd64        2.0.4+dfsg1-2ubunt amd64              Simple DirectMedia Layer
ii  libsdl2-dev                2.0.4+dfsg1-2ubunt amd64              Simple DirectMedia Layer development files
```

**Comment by severach** [comment]
Crtl-Alt-F is working fine here in Arch Linux and Cinnamon.
````
% pacman -Qs | grep sdl
local/lib32-sdl 1.2.15-6
local/sdl 1.2.15-8
local/sdl2 2.0.5-3
local/sdl_image 1.2.12-4
local/sdl_mixer 1.2.12-5
local/sdl_pango 0.1.2-6
local/sdl_ttf 2.0.11-4
````

**Comment by stsp** [comment]
OK, good.
So somehow the problem is specific to my
own setup.

**Comment by stsp** [comment]
There seems to be an SDL bug already filled:
https://bugzilla.libsdl.org/show_bug.cgi?id=1859
I'll try the patch from there.

**Comment by stsp** [comment]
https://github.com/i3/i3/issues/1413

**Comment by severach** [comment]
Spoke too soon. Ctrl-Alt-F always switches between full and window. Sometimes when it switches back to a window the full screen font size is applied to the small window. The Fox TUI behaves as if 80x25 is buffered but only 37x11 is visible.

**Comment by stsp** [comment]
Would it be possible to test the patch from an
sdl ticket?

**Comment by severach** [comment]
Yes.

**Comment by stsp** [comment]
So?
https://bugzilla-attachments.libsdl.org/attachment.cgi?id=1715

**Comment by stsp** [comment]
I am confirming that the SDL patch fixes
the full-screen switching for me as well.
Posted the "me too" comment in sdl bugzilla.

@severach will you package dosemu-sdl? :)

@andrewbird can you take some care of this
on the launchpad? I suspect you can reproduce
the problem if you use $_SDL_swrend=(on).

**Comment by stsp** [comment]
Added a launchpad comment:
https://bugs.launchpad.net/ubuntu/+source/libsdl2/+bug/1671954

**Comment by stsp** [comment]
Overall I think this bug is fatal for us.
Lets see if there will be some activity on ubuntu
side about it.

**Comment by severach** [comment]
I packaged vde2 for #327 because I saw no chance that substantial patches from an outside developer would be integrated into 5 year unmaintained software. My package serves as the test to show that the patches don't break functionality for anything else that uses [vde2](https://www.archlinux.org/packages/extra/x86_64/vde2/), qemu and virtualbox.

There is no need for a new package if the sdl2 developers will apply the patch soon. It's also possible that such a simple patch would be accepted by the Arch devs. sdl1 already has many [patches]( https://git.archlinux.org/svntogit/packages.git/tree/trunk?h=packages/sdl). You and others say that it is working in some distros so they must be applying the patch. The more distros using the patch, the more likely that the Arch devs will accept the patch.

dosemu2 is only a git package at this time. Arch users are instructed that git packages have no expectation of actually working. The lack of a patched sdl2 package on Arch is no reason to delay release.

This bug doesn't crash anything and you can hit the key a few more times to correct the font. This bug isn't a show stopper.

**Comment by stsp** [comment]
> This bug doesn't crash anything and you can hit the key a few more times to correct the font.

For me it manifests differently.
It happens when _entering_ fullscreen (not leaving),
and, even though the existing image resizes to full-screen,
the new drawing happens to the small window of the original
size. And no repeated presses make it sane. So its quite bad,
but is not nearly as bad as a non-working tar in fedora-25,
because of which I can't even create the package.

> There is no need for a new package if the sdl2 developers will apply the patch soon.

Its sitting there for many years.

> The more distros using the patch, the more likely that the Arch devs will accept the patch.

Yes, but I wish ubuntu to take it at least.
Anyway, I will not wait for this to happen if
fedora provides a working tar in a mean time.

**Comment by stsp** [comment]
Is anyone interested to test the fixed
SDL package in ubuntu? They say it
will be removed if not tested.
I only have fedora at home, so a bit
problematic to test.
https://bugs.launchpad.net/ubuntu/+source/libsdl2/+bug/1679573
@jharrison022 maybe you can?
You reported that in #437.

**Comment by severach** [comment]
Looks good to me. The patch is present.
````
% pacman -Q sdl2
sdl2 2.0.7-1
````

**Comment by stsp** [comment]
> % pacman -Q sdl2

They ask to test exactly the experimental ubuntu
package.

**Comment by stuaxo** [comment]
I'll try this, on Ubuntu 17.10

**Comment by stsp** [comment]
Thanks!
You can mark the ubuntu bug as
affecting you and then post the testing
results.

**Comment by stuaxo** [comment]
I'm having trouble reproducing the problem.  - this is on artful 64 bit.

I ran ```dosemu -S``` then full screened and back again a few times.

Maybe this distro is too new

```
$  sudo apt search libsdl2
Sorting... Done
Full Text Search... Done
libsdl2-2.0-0/artful,now 2.0.6+dfsg1-3ubuntu1 amd64 [installed]
  Simple DirectMedia Layer
```

I've got an older laptop somewhere that might have zesty on it, I'll try there.

**Comment by stuaxo** [comment]
OK, dragged out the other one with zesty... trying to build with SDL2

I installed libsd2 and libsdl2-dev, do I need any other packages - something is not quite right when it tries to detect SDL2
```
 ./configure 
checking whether make supports nested variables... yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to enable C11 features... none needed
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking minix/config.h usability... no
checking minix/config.h presence... no
checking for minix/config.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking for a BSD-compatible install... /usr/bin/install -c
checking for flex... flex
checking lex output file root... lex.yy
checking lex library... -lfl
checking whether yytext is a pointer... yes
checking whether ln -s works... yes
checking whether make sets $(MAKE)... yes
checking for ranlib... ranlib
checking for bison... bison -y
checking for grep that handles long lines and -e... (cached) /bin/grep
checking for egrep... (cached) /bin/grep -E
checking for a sed that does not truncate output... /bin/sed
checking for perl... /usr/bin/perl
checking for special C compiler options needed for large files... no
checking for _FILE_OFFSET_BITS value needed for large files... no
checking for static_assert support... yes
checking for shm_open in -lrt... yes
checking for pow in -lm... yes
configure: Using dynamically loaded plugins...
checking for dlopen in -ldl... yes
checking for __attribute__((optimize("no-stack-protector")))... supported
checking for pthread_create in -lpthread... yes
checking for pthread_getname_np... yes
checking for pthread_setname_np... yes
checking for struct _libc_fpstate... yes
checking for struct _fpstate... yes
checking X11/XKBlib.h usability... yes
checking X11/XKBlib.h presence... yes
checking for X11/XKBlib.h... yes
checking netipx/ipx.h usability... yes
checking netipx/ipx.h presence... yes
checking for netipx/ipx.h... yes
checking linux/ipx.h usability... yes
checking linux/ipx.h presence... yes
checking for linux/ipx.h... yes
configure: EXPERIMENTAL stuff disabled...
configure: Creating plug-in hooks...
=== configuring in libao
./configure: line 1678: syntax error near unexpected token `,'
./configure: line 1678: `XIPH_PATH_AO(, as_fn_error $? "libao development libs not found" "$LINENO" 5)'
Configuration for libao failed, disabling
=== configuring in gpm
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for Gpm_Open in -lgpm... no
configure: error: gpm development libs not found
Configuration for gpm failed, disabling
=== configuring in alsa
./configure: line 1678: syntax error near unexpected token `0,,'
./configure: line 1678: `AM_PATH_ALSA(0,,'
Configuration for alsa failed, disabling
=== configuring in ladspa
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... ^[[5~none needed
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking ladspa.h usability... no
checking ladspa.h presence... no
checking for ladspa.h... no
configure: error: *** ladspa not found!
Configuration for ladspa failed, disabling
=== configuring in term
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for SLANG... no
configure: error: slang development headers not found
Configuration for term failed, disabling
=== configuring in X
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for X11... yes
configure: Compiling with X support...
checking for XEXT... yes
checking for fc-cache... /usr/bin/fc-cache
checking for mkfontdir... /usr/bin/mkfontdir
checking for bdftopcf... /usr/bin/bdftopcf
checking font compression method... gzip
checking for gzip... /bin/gzip
configure: creating ./config.status
config.status: creating Makefile.conf
config.status: creating _x_config.h
config.status: _x_config.h is unchanged
=== configuring in sdl
./configure: line 1676: syntax error near unexpected token `$SDL_VERSION,,'
./configure: line 1676: `AM_PATH_SDL2($SDL_VERSION,,'
Configuration for sdl failed, disabling
=== configuring in fluidsynth
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for FLUS... no
configure: error: fluidsynth development libs not found
Configuration for fluidsynth failed, disabling
=== configuring in munt
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for mt32emu_create_context in -lmt32emu... no
configure: error: munt development libs not found
Configuration for munt failed, disabling
=== configuring in sdl1
./configure: line 1675: syntax error near unexpected token `$SDL_VERSION,,'
./configure: line 1675: `AM_PATH_SDL($SDL_VERSION,,'
Configuration for sdl1 failed, disabling
configure: Compiling without debug info...
configure: Compiling with default target CPU...
checking for gcc actually being clang... no
configure: Compiling as PIE for x86_64...
configure: creating ./config.status
config.status: creating Makefile.conf
config.status: creating dosemu2.spec
config.status: creating man/dosemu.1
config.status: creating man/dosemu.bin.1
config.status: creating man/ru/dosemu.1
config.status: creating man/ru/dosemu.bin.1
config.status: creating src/include/config.h
config.status: executing Makefile commands
checking whether ln -s works... yes

Enabling plugins: 
	extra_charsets 
	X 
	midimisc 
	charsets 
	console 
	modemu

configure: WARNING: No audio plugins available!

```

**Comment by stsp** [comment]
I added the missing sdl.m4 file to avoid
such problem for sdl1. But sdl2 seems
to have all the needed file so I don't know
why this happens.
You probably need to run `autogen.sh`
again after installing SDL. See if that helps.

**Comment by stuaxo** [comment]
It only happened on the computer with the older zesty distro.

I tried running autogen on it, and it looks like its missing it...  
- I already installed all the libsdl-* packages.

$ ./autogen.sh 
Generating toplevel configure script...
Generating configure script for src/plugin/X...
 Generating configure script for src/plugin/alsa...
Generating configure script for src/plugin/fdpp...
Generating configure script for src/plugin/fluidsynth...
Generating configure script for src/plugin/gpm...
Generating configure script for src/plugin/ladspa...
Generating configure script for src/plugin/libao...
Generating configure script for src/plugin/munt...
Generating configure script for src/plugin/sdl...
Generating configure script for src/plugin/sdl1...
configure.ac:5: warning: macro 'AM_PATH_SDL' not found in library
configure.ac:5: error: possibly undefined macro: AM_PATH_SDL
      If this token and others are legitimate, please use m4_pattern_allow.
      See the Autoconf documentation.
autoreconf: /usr/bin/autoconf failed with exit status: 1
Warning: failure generating configure for src/plugin/sdl1
Generating configure script for src/plugin/svgalib...
Generating configure script for src/plugin/term...
Generating configure script for src/plugin/vde...

**Comment by stsp** [comment]
This problem is already fixed, please update.

**Comment by stuaxo** [comment]
Haven't had much luck reproducing it.  On zesty I installed SDL with apt and got SDL: 2.0.1+dfsg1-1.

I couldnt repro the bug using freedos EDIT in 25 and 43 line modes.


I also tried: 16.04 in virtualbox - it installed libsdl 2.0.14 dfsg  but couldn't start dosbox

**Comment by stsp** [comment]
Thanks for trying.
Just how much Ctrl-Alf-f switches have
you done? It may not happen from the
first try. 20-30 switches should be enough
to say you don't have this bug.

**Comment by stuaxo** [comment]
I just retried - hitting Ctrl-Alt-F for about as fast as I can for about 30 seconds.   Maybe Freedos EDIT was the wrong thing for me to test in and something else is going on.

Normally, I'm pretty good at breaking software on Ubuntu :) so a bit annoying I didn't succeed here.

Outside of SDL2, the behaviour of fullscreen with the X driver wasn't right - when going to fullscreen the second time, it's the right size, but the window is offset by the height of the the gnome menu bar.

https://github.com/stsp/dosemu2/issues/567

I also noticed when you click the mouse in EDIT, the mouse starts flickering really fast and it's a bit hard to interact with EDIT itself.
